### ISSUE TITLE
feat(`forge test`) Verbose JSON output.

### ISSUE THREAD

**Issue by bowd** [foundry-rs/foundry#8735]
### Component

Forge

### Describe the feature you would like

I've been playing around with this idea of building an interactive testing environment for `forge`. 

I find my self a lot of the time copy-pasting long test names in order to narrow down when I'm debugging or refactoring. I'd like a bit of UI that helps me by keeping track of failed tests and letting me dive in and out of the filtering options. I'd also like auto toggle verbosity on when I'm running a single test. And in the verbose mode have more control over the trace, collapsing longer sections, prettier structs, filtering some vm related logs, toggling setUp, toggling all calls to a certain function, etc. I think working with the traces is amazing but the UX can be improved a lot. I find this especially true when doing more complex fork test scenarios where there's a lot of stuff going on.

I think I can build this in two ways:

1. As stand-alone wrapper over the `forge test` command.
2. As an internal tool in `forge`.

This feature requests is in line with (1). In order to be able to build this with some quality of life, I would need a way to parse the traces. Either `--json` output would include traces, and follow the same rules around verbosity wrt what it includes in the object, or if there are backward compatibility worries, a `--json-detailed` command which would include the traces.

But, apart from this I'm also curious if (2) would be interesting for the maintainers. I'm sure it wouldn't be a priority in any way right now, but I'm curious if you find it interesting. The advantage of (2) is that you could do some more interesting things, like including chisel. And I've been toying with this idea of a sort of "debugger" where the system would integrate chisel and the forge test runner, and maybe there could be a cheatcode like `vm.debug(abi.encode(...localVars))` and what your chisel's underlying REPL contract actually executes the test and when it hits the `vm.debug` statement that becomes a delegatecall back to the REPL contract's run function generated to receive the localVars that you have passed in. I think it's doable but it requires a bunch of moving pieces and a decent amount of AST parsing. 

Curious to hear thoughts on the viability of the detailed `--json` output, and also if y'all find (2) interesting.

**Comment by zerosnacks** [comment]
Hi @bowd, regarding (1): I think `--json` should consistently adhere to verbosity rules. If you could point to cases where this is not the case this would be valuable. For (2): I would like to avoid adding additional tools due to the maintainance overhead but rather make modifications to existing tools to make sure they are user friendly. The best way for (2) is to split your proposal into individual tickets so they can be discussed in more detail. Generally avoid collapsing multiple diverging topics into a single ticket.

**Comment by bowd** [comment]
Ok, you're right, focusing on `--json` output, basically I'd love it if the output would also contain the call-traces and console.logs that the normal output contains as well.
Currently `--json` output's structure doesn't change if I increase verbosity 
```json
{
  "test/oracles/ChainlinkRelayerFactory.t.sol:ChainlinkRelayerFactoryTest_deployRelayer": {
    "duration": "1ms 893us 708ns",
    "test_results": {
      "test_revertsWhenDeployingToAddressWithCode()": {
        "status": "Success",
        "reason": null,
        "counterexample": null,
        "logs": [
          { ... },
          { ... },
        ],
        "kind": {
          "Unit": {
            "gas": 104634
          }
        },
        "labeled_addresses": { ... },
        "duration": {
          "secs": 0,
          "nanos": 277375
        },
        "breakpoints": {}
      }
    },
    "warnings": []
  }
}
```

It would be extremely useful if we could also include the `callTraces` and `console.log` output that the `-vvvv` verbosity provides as well.

**Comment by bowd** [comment]
You're right, updated the issue description focusing on the `--json` output. Thanks!

**Comment by ShantelPeters** [comment]
Hi @bowd please can i be assigned to this issue

**Comment by zerosnacks** [comment]
Related: https://github.com/foundry-rs/foundry/issues/8789#issuecomment-2325892337

**Comment by snehendu098** [comment]
Is this issue still relevant

**Comment by JdmKash** [comment]
Hi — I'd like to take this on if it's still open (looks unassigned since the Feb 2025 unassignment).
To confirm scope before I start: the goal is that forge test --json output should respect verbosity flags the way the standard output does — i.e. at higher verbosity (-vvv+) the JSON should include call traces, and console.log/emitted logs should be represented consistently, rather than the JSON structure staying fixed regardless of -v level. Per @zerosnacks' earlier comment, this is about making --json consistently adhere to verbosity rules.
A couple of scope questions so I build the right thing:

Should traces be added as a structured field on the existing JSON test-result object (alongside logs, kind, etc.), or is there a preferred shape? I want the JSON to stay machine-parseable rather than embedding the human-formatted trace string.
Is the intended behavior to mirror exactly what verbosity levels gate in standard output (e.g. traces at -vvv, setup traces at -vvvv), or just to stop ignoring verbosity entirely?

Happy to align on the output shape first since #8794 tracks output-mode consistency broadly. Thanks!

**Comment by mattsse** [comment]
Closing as fixed. #9034 made verbose JSON test output include traces, and #9074 added decoded logs. Current forge test --json output now responds to verbosity as requested.
